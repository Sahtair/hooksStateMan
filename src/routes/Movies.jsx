import React from 'react';
import styled from 'styled-components';
import {Link} from 'react-router-dom';

import {MoviesContextProvider, useMoviesState} from '../context/movies';
import {useMoviesActions} from '../context/movies/useMoviesActions';
import {
  SET_MOVIES_LOADING,
  SET_MOVIES_LIST,
  SET_MOVIES_ERROR,
} from '../context/movies/actionTypes';

import {withContext} from '../components/withContext';
import {Loader} from '../components/loader';

const MoviesStyled = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  flex: 1;
  background-color: #cca57d;

  .title {
    margin: 32px 0;
    font-size: 64px;
    color: #563421;
  }
`;

const MoviesListStyled = styled.ul`
  list-style: none;
  margin: 0;
  padding: 0;
`;

const MovieStyled = styled.li`
  margin-bottom: 8px;
  padding: 8px;
  border-radius: 8px;
  border: 2px solid #563421;
  background-color: #9a835e;

  :hover {
    background-color: #96705f;
  }

  .link {
    text-decoration: none;
    color: inherit;
  }
`;

const MoviesView = () => {
  console.log(useMoviesState, useMoviesState())
  const {
    state: {
      loading, movies,
    },
  } = useMoviesState();
  console.log(movies)
  const {fetchMovies} = useMoviesActions();

  React.useEffect(() => {
    fetchMovies();
  }, []);

  return (
    <MoviesStyled>
      <div className="title">Movies</div>
      {
      loading ?
      <Loader /> :
      (
      <MoviesListStyled>
        {
          movies.map((movie) => (
            <MovieStyled key={movie._id}>
              <Link className="link" data-testid={`link-to-${movie._id}`} to={`/movies/${movie._id}`}>
                {movie.name}
              </Link>
            </MovieStyled>
          ))
        }
      </MoviesListStyled>
      )
      }
    </MoviesStyled>
  );
};

export const Movies = withContext(MoviesContextProvider)(MoviesView);

