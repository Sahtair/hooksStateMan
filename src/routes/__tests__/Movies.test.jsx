import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import {fireEvent, render} from "@testing-library/react";

import { Movies } from "../Movies";

jest.mock("react-router-dom", () => {
  return {
    Link: ({ to, className }) => {
      return <a href={to} className={className} />;
    },
  };
});

jest.mock("../../context/movies", () => {
  return {
    ...jest.requireActual("../../context/movies"),
    useMoviesState: () => {
      return {
        state: {
          loading: false,
          movies: [{ _id: 1, name: "Movie1" },
          { _id: 2, name: "Movie2" },
          { _id: 3, name: "Movie3" },],
          error: null,
        },
      };
    },
  };
})

describe(`Tests for ${Movies.name}`, () => {


  beforeEach(jest.clearAllMocks)
  test("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Movies />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test("Initial render snapshot", () => {
    const component = renderer.create(<Movies />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("When clicking on a link, the link is clicked", () => {
    const {getAllByRole} = render(<Movies />);

    const link = getAllByRole("link")[0];
    fireEvent.click(link);

    expect(link).toBeVisible();
    expect(link.href).toContain("/movies/1");
  })
});
