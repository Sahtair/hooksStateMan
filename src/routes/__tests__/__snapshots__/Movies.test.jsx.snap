// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for WithContext Initial render snapshot 1`] = `
<div
  className="sc-gsDKAQ bzYqYC"
>
  <div
    className="title"
  >
    Movies
  </div>
  <ul
    className="sc-dkPtRN gQguWw"
  >
    <li
      className="sc-hKwDye lbcwmk"
    >
      <a
        className="link"
        href="/movies/1"
      />
    </li>
    <li
      className="sc-hKwDye lbcwmk"
    >
      <a
        className="link"
        href="/movies/2"
      />
    </li>
    <li
      className="sc-hKwDye lbcwmk"
    >
      <a
        className="link"
        href="/movies/3"
      />
    </li>
  </ul>
</div>
`;
